File: android-beans/src/test/java/rocks/frieler/android/beans/BeanRegistryTest.java
package rocks.frieler.android.beans;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.List;

import rocks.frieler.android.beans.scopes.ScopedFactoryBean;
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler;
import rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBean;
import rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBean;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBean.prototype;
import static rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBean.lazy;

@RunWith(MockitoJUnitRunner.class)
public class BeanRegistryTest {
    private BeanRegistry beanRegistry = new BeanRegistry();

    @Test
    public void testLookUpBeanReturnsNullWithoutABeanWithTheDesiredName() {
        Object bean = beanRegistry.lookUpBean("bean", Object.class);

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanReturnsNullWhenTheBeanWithTheDesiredNameIsOfAnIncompatibleType() {
        String name = "bean";
        Object registeredBean = new Object();

        beanRegistry.registerBean(name, registeredBean);
        BeanRegistryTest bean = beanRegistry.lookUpBean(name, BeanRegistryTest.class);

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanReturnsSingletonBeanWithTheDesiredNameAndType() {
        String name = "bean";
        Object registeredBean = new Object();

        beanRegistry.registerBean(name, registeredBean);
        Object bean = beanRegistry.lookUpBean(name, Object.class);

        assertThat(bean, is(registeredBean));
    }

    @Test
    public void testLookUpBeanByTypeReturnsNullWithoutMatchingBean() {
        beanRegistry.registerBean("bean", new Object());
        BeanRegistryTest bean = beanRegistry.lookUpBean(BeanRegistryTest.class);

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanByTypeReturnsSingleMatchingBean() {
        Object registeredBean = new Object();

        beanRegistry.registerBean("bean", registeredBean);
        Object bean = beanRegistry.lookUpBean(Object.class);

        assertThat(bean, is(registeredBean));
    }

    @Test
    public void testLookUpBeanByTypeReturnsFirstMatchingBean() {
        beanRegistry.registerBean("bean1", new Object());
        beanRegistry.registerBean("bean2", new Object());
        Object bean = beanRegistry.lookUpBean(Object.class);

        assertThat(bean, is(notNullValue()));
    }

    @Test
    public void testLookUpBeansByTypeReturnsAllBeansAssignableToTheDesiredType() {
        beanRegistry.registerBean("object", new Object());
        beanRegistry.registerBean("long", 42L);
        beanRegistry.registerBean("double", 3.14);

        List<Number> numbers = beanRegistry.lookUpBeans(Number.class);

        assertThat(numbers.size(), is(2));
        assertThat(numbers.contains(42L), is(true));
        assertThat(numbers.contains(3.14), is(true));
    }

    @Test
    public void testRegisteredBeanWithoutExplicitNameIsRegisteredWithGeneratedName() {
        Object bean = new Object();

        beanRegistry.registerBean(bean);

        assertThat(beanRegistry.lookUpBean(Object.class.getName(), Object.class), is(sameInstance(bean)));
    }

    @Test
    public void testGeneratedBeanNameUsesAscendingNumbersToInCaseOfCollisions() {
        Object bean1 = new Object();
        Object bean2 = new Object();

        beanRegistry.registerBean(bean1);
        beanRegistry.registerBean(bean2);

        assertThat(beanRegistry.lookUpBean(Object.class.getName(), Object.class), is(sameInstance(bean1)));
        assertThat(beanRegistry.lookUpBean(Object.class.getName() + "2", Object.class), is(sameInstance(bean2)));
    }

    /* tests for bean-scopes: */
    @Mock private ScopedFactoryBeanHandler beanScope;

    @Before
    public void setupBeanScope() {
        when(beanScope.getName()).thenReturn("aScope");
        beanRegistry.addBeanScope(beanScope);
    }

    @Test
    public void testLookUpBeanReturnsNullWhenScopedFactoryBeanWithDesiredNameProducesWrongType() {
        final String name = "bean";
        @SuppressWarnings("unchecked") ScopedFactoryBean<Object> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(Object.class);

        beanRegistry.registerBean(name, factoryBean);
        Object bean = beanRegistry.lookUpBean(name, BeanRegistryTest.class);

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanReturnsNullWithoutSuitableScopeForScopedFactoryBean() {
        final String name = "bean";
        @SuppressWarnings("unchecked") ScopedFactoryBean<BeanRegistryTest> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(BeanRegistryTest.class);
        when(factoryBean.getScope()).thenReturn("otherScope");

        beanRegistry.registerBean(name, factoryBean);
        BeanRegistryTest bean = beanRegistry.lookUpBean(name, factoryBean.getBeanType());

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanReturnsNullDespiteScopedFactoryBeanWithDesiredNameWhenScopeIsNotActive() {
        when(beanScope.isActive()).thenReturn(false);
        final String name = "bean";
        final String scope = beanScope.getName();
        @SuppressWarnings("unchecked") ScopedFactoryBean<BeanRegistryTest> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(BeanRegistryTest.class);
        when(factoryBean.getScope()).thenReturn(scope);

        beanRegistry.registerBean(name, factoryBean);
        BeanRegistryTest bean = beanRegistry.lookUpBean(name, factoryBean.getBeanType());

        assertThat(bean, is(nullValue()));
    }

    @Test
    public void testLookUpBeanReturnsScopedBeanFromActiveScopeWithTheDesiredNameAndType() {
        when(beanScope.isActive()).thenReturn(true);
        final String name = "bean";
        final String scope = beanScope.getName();
        final Object scopedBean = new Object();
        @SuppressWarnings("unchecked") ScopedFactoryBean<Object> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(Object.class);
        when(factoryBean.getScope()).thenReturn(scope);
        when(factoryBean.produceBean()).thenReturn(scopedBean);
        doAnswer(invocation -> ((ScopedFactoryBean) invocation.getArgument(1)).produceBean())
                .when(beanScope).getBean(eq(name), any(ScopedFactoryBean.class));

        beanRegistry.registerBean(name, factoryBean);
        Object bean = beanRegistry.lookUpBean(name, factoryBean.getBeanType());

        assertThat(bean, is(sameInstance(scopedBean)));
    }

    @Test
    public void testLookUpBeanByTypeReturnsMatchingBeanFromScopedFactoryBean() {
        when(beanScope.isActive()).thenReturn(true);
        final String name = "bean";
        final String scope = beanScope.getName();
        @SuppressWarnings("unchecked") ScopedFactoryBean<BeanRegistryTest> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(BeanRegistryTest.class);
        when(factoryBean.getScope()).thenReturn(scope);
        when(factoryBean.produceBean()).thenReturn(this);
        doAnswer(invocation -> ((ScopedFactoryBean) invocation.getArgument(1)).produceBean())
                .when(beanScope).getBean(eq(name), any(ScopedFactoryBean.class));

        beanRegistry.registerBean(name, factoryBean);
        BeanRegistryTest bean = beanRegistry.lookUpBean(factoryBean.getBeanType());

        assertThat(bean, is(this));
    }

    @Test
    public void testLookUpBeansByTypeReturnsAllBeansAssignableToTheDesiredTypeIncludingScopedBeans() {
        when(beanScope.isActive()).thenReturn(true);
        beanRegistry.registerBean("object", new Object());
        beanRegistry.registerBean("long", 42L);
        final String name = "double";
        final String scope = beanScope.getName();
        @SuppressWarnings("unchecked") ScopedFactoryBean<Double> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(Double.class);
        when(factoryBean.getScope()).thenReturn(scope);
        when(factoryBean.produceBean()).thenReturn(3.14);
        doAnswer(invocation -> ((ScopedFactoryBean) invocation.getArgument(1)).produceBean())
                .when(beanScope).getBean(eq(name), any(ScopedFactoryBean.class));
        beanRegistry.registerBean(name, factoryBean);

        List<Number> numbers = beanRegistry.lookUpBeans(Number.class);

        assertThat(numbers.size(), is(2));
        assertThat(numbers.contains(42L), is(true));
        assertThat(numbers.contains(3.14), is(true));
    }

    @Test
    public void testGeneratedBeanNameForScopedFactoryBeansIsDerivedFromProducedBeanType() {
        final ScopedFactoryBeanHandler scope = mock(ScopedFactoryBeanHandler.class);
        when(scope.getName()).thenReturn("scope");
        when(scope.isActive()).thenReturn(true);
        beanRegistry.addBeanScope(scope);

        @SuppressWarnings("unchecked") ScopedFactoryBean<BeanRegistryTest> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getScope()).thenReturn("scope");
        when(factoryBean.getBeanType()).thenReturn(BeanRegistryTest.class);
        //noinspection unchecked
        when(scope.<BeanRegistryTest>getBean(eq(BeanRegistryTest.class.getName()), any(ScopedFactoryBean.class))).thenReturn(this);

        beanRegistry.registerBean(factoryBean);

        assertThat(beanRegistry.lookUpBean(this.getClass().getName(), this.getClass()), is(sameInstance(this)));
    }

    @Test
    public void testBeanRegistrySupportsLazyInstantiationInSingletonScope() {
        final String name = "lazyInstantiatedBean";
        final SingletonScopedFactoryBean<BeanRegistryTest> singletonFactory = lazy(BeanRegistryTest.class, () -> this);

        beanRegistry.registerBean(name, singletonFactory);
        final BeanRegistryTest beanInstance = beanRegistry.lookUpBean(name, BeanRegistryTest.class);

        assertThat(beanInstance, is(singletonFactory.produceBean()));
    }

    @Test
    public void testBeanRegistrySupportsPrototypeScope() {
        final String name = "prototypeBean";
        final PrototypeScopedFactoryBean<BeanRegistryTest> prototype = prototype(BeanRegistryTest.class, () -> this);

        beanRegistry.registerBean(name, prototype);
        final BeanRegistryTest beanInstance = beanRegistry.lookUpBean(name, BeanRegistryTest.class);

        assertThat(beanInstance, is(prototype.produceBean()));
    }

    /* tests for post-processing: */
    @Mock private BeanPostProcessor beanPostProcessor;

    @Test
    public void testRegisteredBeanPostProcessorGetsNewBeansToPostProcess() {
        Object originalBean = new Object();
        Object replacementBean = new Object();
        when(beanPostProcessor.postProcessBean("bean", originalBean)).thenReturn(replacementBean);

        beanRegistry.registerBeanPostProcessor(beanPostProcessor);
        beanRegistry.registerBean("bean", originalBean);

        verify(beanPostProcessor).postProcessBean("bean", originalBean);
        assertThat(beanRegistry.lookUpBean("bean", Object.class), is(replacementBean));
    }

    @Test
    public void testRegisteredBeanPostProcessorGetsExistingBeansToPostProcess() {
        Object originalBean = new Object();
        Object replacementBean = new Object();
        when(beanPostProcessor.postProcessBean("bean", originalBean)).thenReturn(replacementBean);

        beanRegistry.registerBean("bean", originalBean);
        beanRegistry.registerBeanPostProcessor(beanPostProcessor);

        verify(beanPostProcessor).postProcessBean("bean", originalBean);
        assertThat(beanRegistry.lookUpBean("bean", Object.class), is(replacementBean));
    }

    @Test
    public void testRegisteredBeanPostProcessorGetsScopedBeanToPostProcessWhenItGetsProduced() {
        when(beanScope.isActive()).thenReturn(true);
        final String name = "bean";
        final String scope = beanScope.getName();
        final Object originalBean = new Object();
        @SuppressWarnings("unchecked") ScopedFactoryBean<Object> factoryBean = mock(ScopedFactoryBean.class);
        when(factoryBean.getBeanType()).thenReturn(Object.class);
        when(factoryBean.getScope()).thenReturn(scope);
        when(factoryBean.produceBean()).thenReturn(originalBean);
        doAnswer(invocation -> ((ScopedFactoryBean) invocation.getArgument(1)).produceBean())
                .when(beanScope).getBean(eq(name), any(ScopedFactoryBean.class));

        when(beanPostProcessor.postProcessBean(name, factoryBean)).thenReturn(factoryBean);
        Object replacementBean = new Object();
        when(beanPostProcessor.postProcessBean(name, originalBean)).thenReturn(replacementBean);

        beanRegistry.registerBean(name, factoryBean);
        beanRegistry.registerBeanPostProcessor(beanPostProcessor);

        assertThat(beanRegistry.lookUpBean(name, Object.class), is(replacementBean));
        verify(beanPostProcessor).postProcessBean(name, originalBean);
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/BeanRegistryTest.kt
package rocks.frieler.android.beans

import assertk.assertThat
import assertk.assertions.containsOnly
import assertk.assertions.isEqualTo
import assertk.assertions.isIn
import assertk.assertions.isNotEqualTo
import assertk.assertions.isNull
import assertk.assertions.isSameAs
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import rocks.frieler.android.beans.scopes.ScopedFactoryBean
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler
import rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBean
import rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBean
import kotlin.reflect.KClass
import kotlin.reflect.jvm.jvmName

class BeanRegistryTest {
    private val beanRegistry = BeanRegistry()

    private val parent = mock<BeansProvider>().apply {
        whenever(lookUpOptionalBean(any(), any<KClass<*>>())).thenReturn(null)
        whenever(lookUpOptionalBean(any<KClass<*>>())).thenReturn(null)
        whenever(lookUpBeans(any<KClass<*>>())).thenReturn(emptyList())
    }
    private val beanRegistryWithParent = BeanRegistry(parent)

    @Test
    fun `lookUpOptionalBean() by name and type returns null without a bean with the desired name`() {
        val bean = beanRegistry.lookUpOptionalBean("bean", Any::class)

        assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns null when the bean with the desired name is of an incompatible type`() {
        val name = "bean"
        val registeredBean = Any()

        beanRegistry.registerBean(name, registeredBean)
        val bean = beanRegistry.lookUpOptionalBean(name, BeanRegistryTest::class)

		assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns registered bean with the desired name and type`() {
        val name = "bean"
        val registeredBean = Any()

        beanRegistry.registerBean(name, registeredBean)
        val bean = beanRegistry.lookUpOptionalBean(name, Any::class)

        assertThat(bean).isSameAs(registeredBean)
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns bean with the desired name and type from parent if no local one matches`() {
        val name = "bean"
        val inheritedBean = Any()
        whenever(parent.lookUpOptionalBean(name, Any::class)).thenReturn(inheritedBean)

        val bean = beanRegistryWithParent.lookUpOptionalBean(name, Any::class)

        assertThat(bean).isSameAs(inheritedBean)
    }

    @Test
    fun `lookUpOptionalBean() by type returns null without a matching bean`() {
        beanRegistry.registerBean("bean", Any())
        val bean = beanRegistry.lookUpOptionalBean(BeanRegistryTest::class)

		assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by type returns single matching bean`() {
        val registeredBean = Any()

        beanRegistry.registerBean("bean", registeredBean)
        val bean = beanRegistry.lookUpOptionalBean(Any::class)

		assertThat(bean).isSameAs(registeredBean)
    }

    @Test
    fun `lookUpOptionalBean() by type returns one of multiple matching beans`() {
		val registeredBean1 = Any()
		val registeredBean2 = Any()

		beanRegistry.registerBean("bean1", registeredBean1)
		beanRegistry.registerBean("bean2", registeredBean2)
        val bean = beanRegistry.lookUpOptionalBean(Any::class)

		assertThat(bean).isIn(registeredBean1, registeredBean2)
    }

    @Test
    fun `lookUpOptionalBean() by type returns bean with the desired type from parent if no local one matches`() {
        val inheritedBean = Any()
        whenever(parent.lookUpOptionalBean(Any::class)).thenReturn(inheritedBean)

        val bean = beanRegistryWithParent.lookUpOptionalBean(Any::class)

        assertThat(bean).isSameAs(inheritedBean)
    }

    @Test
    fun `lookUpBeans() by type returns all beans assignable to the desired type`() {
        beanRegistry.registerBean("object", Any())
        beanRegistry.registerBean("long", 42L)
        beanRegistry.registerBean("double", 3.14)
        val numbers = beanRegistry.lookUpBeans(Number::class)

		assertThat(numbers).containsOnly(42L, 3.14)
    }

    @Test
    fun `lookUpBeans() by type returns all registered and inherited beans`() {
        val inheritedBean = Any()
        whenever(parent.lookUpBeans(Any::class)).thenReturn(listOf(inheritedBean))
        val registeredBean = Any()
        beanRegistryWithParent.registerBean(registeredBean)

        val beans = beanRegistryWithParent.lookUpBeans(Any::class)

        assertThat(beans).containsOnly(registeredBean, inheritedBean)
    }

    @Test
    fun `registerBean() without explicit name generates a name from the bean-type`() {
        val bean = Any()

        val generatedName = beanRegistry.registerBean(bean)

        assertThat(beanRegistry.lookUpBean(generatedName, Any::class)).isSameAs(bean)
    }

    @Test
    fun `generated bean names avoid collisions`() {
        val bean1 = Any()
        val bean2 = Any()

        val generatedNameForBean1 = beanRegistry.registerBean(bean1)
        val generatedNameForBean2 = beanRegistry.registerBean(bean2)

        assertThat(generatedNameForBean2).isNotEqualTo(generatedNameForBean1)
        assertThat(beanRegistry.lookUpBean(generatedNameForBean1, Any::class)).isSameAs(bean1)
		assertThat(beanRegistry.lookUpBean(generatedNameForBean2, Any::class)).isSameAs(bean2)
    }

    /* tests for bean-scopes: */
    private val beanScope: ScopedFactoryBeanHandler = mock()

    @BeforeEach
    fun setupBeanScope() {
        whenever(beanScope.name).thenReturn("aScope")
        beanRegistry.addBeanScope(beanScope)
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns null when the ScopedFactoryBean with desired name produces the wrong type`() {
        val name = "bean"
        val factoryBean: ScopedFactoryBean<Any> = mock()
        whenever(factoryBean.beanType).thenReturn(Any::class)

        beanRegistry.registerBean(name, factoryBean)
        val bean: Any? = beanRegistry.lookUpOptionalBean(name, BeanRegistryTest::class)

        assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns null without a suitable scope for a matching ScopedFactoryBean`() {
        val name = "bean"
        val factoryBean: ScopedFactoryBean<BeanRegistryTest> = mock()
        whenever(factoryBean.beanType).thenReturn(BeanRegistryTest::class)
        whenever(factoryBean.scope).thenReturn("otherScope")

        beanRegistry.registerBean(name, factoryBean)
        val bean = beanRegistry.lookUpOptionalBean(name, factoryBean.beanType)

        assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns null when the suitable ScopedFactoryBean is not active`() {
        whenever(beanScope.isActive).thenReturn(false)
        val name = "bean"
        val scope = beanScope.name
        val factoryBean: ScopedFactoryBean<BeanRegistryTest> = mock()
        whenever(factoryBean.beanType).thenReturn(BeanRegistryTest::class)
        whenever(factoryBean.scope).thenReturn(scope)

        beanRegistry.registerBean(name, factoryBean)
        val bean = beanRegistry.lookUpOptionalBean(name, factoryBean.beanType)

        assertThat(bean).isNull()
    }

    @Test
    fun `lookUpOptionalBean() by name and type returns scoped bean from active scope with the desired name and type`() {
        whenever(beanScope.isActive).thenReturn(true)
        val name = "bean"
        val scope = beanScope.name
        val scopedBean = Any()
        val factoryBean: ScopedFactoryBean<Any> = mock()
        whenever(factoryBean.beanType).thenReturn(Any::class)
        whenever(factoryBean.scope).thenReturn(scope)
        whenever(factoryBean.produceBean(beanRegistry)).thenReturn(scopedBean)
		whenever(beanScope.getBean<Any>(eq(name), any(), eq(beanRegistry)))
				.thenAnswer { invocation -> (invocation.getArgument(1) as ScopedFactoryBean<*>).produceBean(beanRegistry) }

        beanRegistry.registerBean(name, factoryBean)
        val bean = beanRegistry.lookUpOptionalBean(name, factoryBean.beanType)

        assertThat(bean).isSameAs(scopedBean)
    }

    @Test
    fun `lookUpOptionalBean() by type returns matching bean from ScopedFactoryBean`() {
        whenever(beanScope.isActive).thenReturn(true)
        val name = "bean"
        val scope = beanScope.name
        val factoryBean: ScopedFactoryBean<BeanRegistryTest> = mock()
        whenever(factoryBean.beanType).thenReturn(BeanRegistryTest::class)
        whenever(factoryBean.scope).thenReturn(scope)
        whenever(factoryBean.produceBean(beanRegistry)).thenReturn(this)
		whenever(beanScope.getBean<Any>(eq(name), any(), eq(beanRegistry)))
				.thenAnswer { invocation -> (invocation.getArgument(1) as ScopedFactoryBean<*>).produceBean(beanRegistry) }

        beanRegistry.registerBean(name, factoryBean)
        val bean = beanRegistry.lookUpOptionalBean(factoryBean.beanType)

        assertThat(bean).isSameAs(this)
    }

    @Test
    fun `lookUpBeans() by type returns all beans assignable to the desired type including scoped beans`() {
        whenever(beanScope.isActive).thenReturn(true)
        beanRegistry.registerBean("object", Any())
        beanRegistry.registerBean("long", 42L)
        val name = "double"
        val scope = beanScope.name
        val factoryBean: ScopedFactoryBean<Double> = mock()
        whenever(factoryBean.beanType).thenReturn(Double::class)
        whenever(factoryBean.scope).thenReturn(scope)
        whenever(factoryBean.produceBean(beanRegistry)).thenReturn(3.14)
		whenever(beanScope.getBean<Any>(eq(name), any(), eq(beanRegistry)))
				.thenAnswer { invocation -> (invocation.getArgument(1) as ScopedFactoryBean<*>).produceBean(beanRegistry) }

        beanRegistry.registerBean(name, factoryBean)
        val numbers = beanRegistry.lookUpBeans(Number::class)

		assertThat(numbers).containsOnly(42L, 3.14)
    }

    @Test
    fun `generated bean name for ScopedFactoryBeans is derived from produced bean-type`() {
        val factoryBean: ScopedFactoryBean<BeanRegistryTest> = mock()
        whenever(factoryBean.beanType).thenReturn(BeanRegistryTest::class)

        val generatedName = beanRegistry.registerBean(factoryBean)

        assertThat(generatedName).isEqualTo(this::class.jvmName)
    }

    @Test
    fun `BeanRegistry supports lazy instantiation in singleton-scope by default`() {
        val name = "lazyInstantiatedBean"
        val singletonFactory = SingletonScopedFactoryBean(BeanRegistryTest::class) { this@BeanRegistryTest }

        beanRegistry.registerBean(name, singletonFactory)
        val beanInstance = beanRegistry.lookUpBean(name, BeanRegistryTest::class)

        assertThat(beanInstance).isSameAs(this)
    }

    @Test
    fun `BeanRegistry supports prototype-scope by default`() {
        val name = "prototypeBean"
        val prototype = PrototypeScopedFactoryBean(BeanRegistryTest::class) { this@BeanRegistryTest }

        beanRegistry.registerBean(name, prototype)
        val beanInstance = beanRegistry.lookUpBean(name, BeanRegistryTest::class)

		assertThat(beanInstance).isSameAs(this)
    }

    /* tests for post-processing: */
    private val beanPostProcessor: BeanPostProcessor = mock()

    @Test
    fun `bean implementing BeanPostProcessor gets detected and receives new beans to post-process`() {
        val originalBean = Any()
        val replacementBean = Any()
        whenever(beanPostProcessor.postProcessBean("bean", originalBean)).thenReturn(replacementBean)

        beanRegistry.registerBean(beanPostProcessor)
        beanRegistry.registerBean("bean", originalBean)

		verify(beanPostProcessor).postProcessBean("bean", originalBean)
        assertThat(beanRegistry.lookUpBean("bean", Any::class)).isSameAs(replacementBean)
    }

    @Test
    fun `bean implementing BeanPostProcessor gets detected and receives existing beans to post-process`() {
        val originalBean = Any()
        val replacementBean = Any()
        whenever(beanPostProcessor.postProcessBean("bean", originalBean)).thenReturn(replacementBean)

        beanRegistry.registerBean("bean", originalBean)
        beanRegistry.registerBean(beanPostProcessor)

		verify(beanPostProcessor).postProcessBean("bean", originalBean)
		assertThat(beanRegistry.lookUpBean("bean", Any::class)).isSameAs(replacementBean)
    }

    @Test
    fun `bean implementing BeanPostProcessor receives scoped bean to post-process when it is produced`() {
        whenever(beanScope.isActive).thenReturn(true)
        val name = "bean"
        val scope = beanScope.name
        val originalBean = Any()
        val factoryBean: ScopedFactoryBean<Any> = mock()
        whenever(factoryBean.beanType).thenReturn(Any::class)
        whenever(factoryBean.scope).thenReturn(scope)
        whenever(factoryBean.produceBean(beanRegistry)).thenReturn(originalBean)
		whenever(beanScope.getBean<Any>(eq(name), any(), eq(beanRegistry)))
				.thenAnswer { invocation -> (invocation.getArgument(1) as ScopedFactoryBean<*>).produceBean(beanRegistry) }
        whenever(beanPostProcessor.postProcessBean(name, factoryBean)).thenReturn(factoryBean)
        val replacementBean = Any()
        whenever(beanPostProcessor.postProcessBean(name, originalBean)).thenReturn(replacementBean)

        beanRegistry.registerBean(name, factoryBean)
        beanRegistry.registerBean(beanPostProcessor)
		val finalBean = beanRegistry.lookUpBean(name, Any::class)

        verify(beanPostProcessor).postProcessBean(name, originalBean)
		assertThat(finalBean).isSameAs(replacementBean)
    }
}
